# Q-1

Download Titanic data from Kaggle. The data in the train.csv file meets your need.

Split your data into training and test sets.

Predict the survival based on the test data you split by creating your model.
Is your model's performance satisfactory? Explain.

Try to improve your model's performance by adding or subtracting some variables.

Explore the advantages and disadvantages of Logistic Regression and discuss with your mentor.

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
import warnings
warnings.filterwarnings(action="ignore")

pd.options.display.max_columns = None

In [2]:
titanic = pd.read_csv("titanic_train.csv")
titanic.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

In [3]:
titanic.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [4]:
titanic.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [5]:
#titanic = titanic[titanic['Cabin'].notna()]
#titanic = titanic[titanic['Age'].notna()]

#titanic

In [6]:
titanic["Sex_"]=pd.get_dummies(titanic["Sex"], drop_first=True)
titanic.groupby(["Embarked"])["PassengerId"].count()

Embarked
C    168
Q     77
S    644
Name: PassengerId, dtype: int64

In [7]:
titanic["EmbarkedC"]=titanic["Embarked"].replace("C",1)
titanic["EmbarkedC"].replace("S",0,inplace=True)
titanic["EmbarkedC"].replace("Q",0,inplace=True)
titanic["EmbarkedS"]=titanic["Embarked"].replace("S",1)
titanic["EmbarkedS"].replace("Q",0,inplace=True)
titanic["EmbarkedS"].replace("C",0,inplace=True)
titanic["EmbarkedQ"]=titanic["Embarked"].replace("Q",1)
titanic["EmbarkedQ"].replace("S",0,inplace=True)
titanic["EmbarkedQ"].replace("C",0,inplace=True)
titanic.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  Sex_  EmbarkedC  \
0      0         A/5 21171   7.2500   NaN        S     1        0.0   
1      0          PC 17599  71.2833   C85        C     0        1.0   
2      0  STON/O2. 3101282   7.9250   NaN        S     0        0.0   
3      0            113803  53.1000  C123        S     0        0.0   
4      0            373450   8.0500   NaN        S     1        0.0   

   EmbarkedS  EmbarkedQ  
0        1.0        0.0  
1        0.0        0.0  
2        1.0        0.0  
3        1.0        0.0  
4        1.0        0.0

In [8]:
titanic.dropna(subset=['Age', 'Embarked'],inplace=True)

In [9]:
Y = titanic["Survived"]

X = titanic[["Pclass","Sex_","Age",
                 "Parch","EmbarkedC","EmbarkedS","EmbarkedQ"]]

X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size = 0.2, random_state = 465)

print("The number of observations in training set is {}".format(X_train.shape[0]))
print("The number of observations in test set is {}".format(X_test.shape[0]))

The number of observations in training set is 569
The number of observations in test set is 143


In [10]:
log_reg = LogisticRegression(solver='liblinear')


In [11]:
log_reg.fit(X_train, y_train)


LogisticRegression(solver='liblinear')

In [12]:
train_accuracy = log_reg.score(X_train, y_train)
test_accuracy = log_reg.score(X_test, y_test)

print('One-vs.-Rest', '-'*30, 
      'Accuracy on Train Data : {:.2f}'.format(train_accuracy), 
      'Accuracy on Test Data  : {:.2f}'.format(test_accuracy), sep='\n')

One-vs.-Rest
------------------------------
Accuracy on Train Data : 0.79
Accuracy on Test Data  : 0.80


In [13]:
log_reg.classes_

array([0, 1], dtype=int64)

In [14]:
y_pred = log_reg.predict(X_test)
y_pred

array([0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0,
       0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0, 0, 1, 0, 0,
       0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 0, 0,
       0, 0, 1, 0, 1, 0, 0, 0, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 1, 0, 0, 0,
       1, 0, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1,
       1, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0,
       0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0], dtype=int64)

In [15]:
log_reg.predict_proba(X_test)
#The first column is the probability of the predicted output being zero
#second is one

array([[0.8771583 , 0.1228417 ],
       [0.70228598, 0.29771402],
       [0.84764819, 0.15235181],
       [0.93183291, 0.06816709],
       [0.3740883 , 0.6259117 ],
       [0.889396  , 0.110604  ],
       [0.68124166, 0.31875834],
       [0.42554585, 0.57445415],
       [0.72096479, 0.27903521],
       [0.95370509, 0.04629491],
       [0.88341646, 0.11658354],
       [0.76617597, 0.23382403],
       [0.58625318, 0.41374682],
       [0.86377396, 0.13622604],
       [0.30099279, 0.69900721],
       [0.81497425, 0.18502575],
       [0.79703096, 0.20296904],
       [0.81497425, 0.18502575],
       [0.8771583 , 0.1228417 ],
       [0.90825217, 0.09174783],
       [0.26222283, 0.73777717],
       [0.86024144, 0.13975856],
       [0.86723098, 0.13276902],
       [0.52149609, 0.47850391],
       [0.45284056, 0.54715944],
       [0.82859922, 0.17140078],
       [0.28725562, 0.71274438],
       [0.08605186, 0.91394814],
       [0.86377396, 0.13622604],
       [0.66158988, 0.33841012],
       [0.

# İmproving Model

In [16]:
titanic["Sib/age"] = titanic["SibSp"]*titanic["Age"]

In [17]:
X_ = titanic[["Pclass","Sex_","Age","Parch","Sib/age"]]
Y_ = titanic["Survived"]




In [18]:
X_train_, X_test_, y_train_, y_test_ = train_test_split(X_, Y_, test_size = 0.2, random_state = 465)

In [19]:
log_reg_ = LogisticRegression()
log_reg_.fit(X_train_, y_train_)

LogisticRegression()

In [20]:
train_accuracy = log_reg_.score(X_train_, y_train_)
test_accuracy = log_reg_.score(X_test_, y_test_)

print('One-vs.-Rest', '-'*30, 
      'Accuracy on Train Data : {:.2f}'.format(train_accuracy), 
      'Accuracy on Test Data  : {:.2f}'.format(test_accuracy), sep='\n')

One-vs.-Rest
------------------------------
Accuracy on Train Data : 0.80
Accuracy on Test Data  : 0.82


With interaction between Sibsp and age, the accuracy of model is increased.

In [23]:
C_values = [0.001,0.01, 0.1,1,10,100, 1000]

accuracy_values = pd.DataFrame(columns=['C_values', 'Train Accuracy', 'Test Accuracy'])

for c in C_values:
    # Apply logistic regression model to training data
    lr = LogisticRegression(penalty = 'l2', C = c, random_state = 0)
    lr.fit(X_train_, y_train_)
    accuracy_values = accuracy_values.append({'C_values': c,
                                              'Train Accuracy': lr.score(X_train_, y_train_),
                                              'Test Accuracy': lr.score(X_test_, y_test_)
                                             }, ignore_index=True)
display(accuracy_values)  

C_values  Train Accuracy  Test Accuracy
0     0.001        0.594025       0.629371
1     0.010        0.766257       0.748252
2     0.100        0.801406       0.825175
3     1.000        0.797891       0.818182
4    10.000        0.796134       0.825175
5   100.000        0.796134       0.825175
6  1000.000        0.796134       0.825175

According to the results, c value was 0.100 and gave the best result.